pyqir-generator/src/python.rs

```diff
@@ -1,174 +1,264 @@
 // Copyright (c) Microsoft Corporation.
 // Licensed under the MIT License.
-use crate::emit::{get_bitcode_base64_string, get_ir_string, write_model_to_file};
-use crate::interop::{
-    ClassicalRegister, Controlled, Instruction, Measured, QuantumRegister, Rotated, SemanticModel,
-    Single,
+
+use crate::{
+    emit,
+    interop::{
+        ClassicalRegister, Controlled, If, Instruction, Measured, QuantumRegister, Rotated,
+        SemanticModel, Single,
+    },
 };
-use log;
-use pyo3::exceptions::PyOSError;
-use pyo3::prelude::*;
+use pyo3::{exceptions::PyOSError, prelude::*};
 
 #[pymodule]
-fn pyqir_generator(_py: Python<'_>, m: &PyModule) -> PyResult<()> {
-    m.add_class::<PyQIR>()?;
-    Ok(())
+fn _native(_py: Python, m: &PyModule) -> PyResult<()> {
+    m.add_class::<SimpleModule>()?;
+    m.add_class::<Qubit>()?;
+    m.add_class::<Ref>()?;
+    m.add_class::<Builder>()?;
+    m.add_class::<BasicQisBuilder>()
 }
 
+const RESULT_NAME: &str = "result";
+const QUBIT_NAME: &str = "qubit";
+
 #[pyclass]
-pub struct PyQIR {
-    pub(super) model: SemanticModel,
+struct SimpleModule {
+    model: SemanticModel,
+    builder: Py<Builder>,
 }
 
 #[pymethods]
-impl PyQIR {
+impl SimpleModule {
     #[new]
-    fn new(name: String) -> Self {
-        PyQIR {
-            model: SemanticModel::new(name),
+    fn new(py: Python, name: String, num_qubits: u64, num_results: u64) -> PyResult<SimpleModule> {
+        let registers = vec![ClassicalRegister::new(RESULT_NAME.to_string(), num_results)];
+
+        let qubits = (0..num_qubits)
+            .map(|i| QuantumRegister::new(QUBIT_NAME.to_string(), i))
+            .collect();
+
+        let model = SemanticModel {
+            name,
+            registers,
+            qubits,
+            instructions: Vec::new(),
+            static_alloc: true,
+        };
+
+        let builder = Py::new(py, Builder::new())?;
+        Ok(SimpleModule { model, builder })
+    }
+
+    #[getter]
+    fn qubits(&self) -> Vec<Qubit> {
```

Comment:
This getter seems to create a new vector each time it's called, which can result in the items at the python level having different hashes. This may be less of a concern for generation than for parsing, since in the latter case the QIR is constant while the former is necessarily mutating it, but it's something I figured I'd call out since it came from a community issue (see https://github.com/qir-alliance/pyqir/issues/39).

Reply:
I implemented `__hash__` and `__richcmp__` for `Qubit` and `Ref`. The hash is now based on the index they represent instead of their object identity and should be consistent across property accesses.

Reply:
Fantastic, thanks!